python/paddle/v2/fluid/layers/nn.py

```diff
@@ -2951,3 +2952,257 @@ def multiplex(inputs, index):
                 'Ids': index},
         outputs={'Out': [out]})
     return out
+
+
+import __init__ as pd
+
+
+class SequenceDecoder:
+    class Cell:
+        modes = ('train', 'decode')
+        kinds = ('step_input', 'state', 'other_shared')
+        dic = {
+            'train': {
+                'step_input': {},
+                'state': {},
+                'others': {},
+            },
+            'decode': {
+                'step_input': {},
+                'state': {},
+                'others': {},
+            }
+        }
+
+        def __init__(self, kind, id, init_var, dtype='float32'):
+            '''
+            kind: state_input, state or other_shared
+            id: identification of this cell.
+            init_var: variable to initialize this cell.
+                init_var can be an Variable or a dic like {'train': var0, 'decode': var1}
+                so that different initialization can be porformed in different mode, this is
+                important when defining states.
+            seqdec: instance of SequenceDecoder.
+            rnn: Dynamic rnn instance.
+            '''
+            self.kind = kind
+            self.id = id
+            self.init_var = init_var
+            self.rnn = None
+            self.dtype = dtype
+            self.mode = None
+
+            assert kind in SequenceDecoder.Cell.kinds
+
+        def set_mode(self, mode, seqdec):
+            self.mode = mode
+            assert self.mode in SequenceDecoder.Cell.modes, \
+            "invalid mode: {}, only {} are valid.".format(
+                self.mode, SequenceDecoder.Cell.modes,)
+            if self.mode == 'train':
+                self.rnn = seqdec.train_rnn
+
+        def create(self):
+            assert self.mode is not None, "mode should be set first by calling `set_mode`"
+            if self.mode == 'train':
+                return self._create_train()
+            else:
+                return self._create_decode()
+
+        def set_updater(self, updater):
```

Comment:
Better to change to `set_states_updater`?

Since updater is a callback, I think `updater_func` or `updater_callback` is more appropriate.